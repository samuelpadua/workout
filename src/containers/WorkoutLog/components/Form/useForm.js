import { useState, useEffect } from 'react'

const maskTimeSpent = value => (
  value
    .toString()
    .replace(/\D/g, '')
    .replace(/\D:/g, '')
    .replace(/^(\d{2})(\d)/g, '$1:$2')
)

const useForm = (callback, validate) => {
  const [values, setValues] = useState({})
  const [errors, setErrors] = useState({ error: false, fields: {} })
  const [isSubmitting, setIsSubmitting] = useState(false)

  useEffect(() => {
    if (Object.keys(errors.fields).length === 0 && isSubmitting) {
      callback()
      setIsSubmitting(false)
      setValues({})
      setErrors({ error: false, fields: {} })
    }
  }, [errors])

  const handleSubmit = (event) => {
    if (event) event.preventDefault()
    setErrors(validate(values))
    setIsSubmitting(true)
  }

  const handleChange = (event) => {
    event.persist()
    let { value } = event.target

    if (event.target.name === 'timeSpent') {
      value = maskTimeSpent(value)
    }

    setValues(_values => ({ ..._values, [event.target.name]: value }))
  }

  return {
    handleChange,
    handleSubmit,
    values,
    errors,
  }
}

export default useForm
