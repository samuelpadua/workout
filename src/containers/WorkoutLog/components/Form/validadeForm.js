const validateTime = (value) => {
  const regexp = /([0-1][0-9]|2[0-4])\:[0-5][0-9]/

  return regexp.test(value)
}

export default function validate(values) {
  const errors = {
    error: false,
    fields: {},
  }

  if (!values.timeSpent) {
    errors.fields.timeSpent = 'Coloque o tempo'
  } else if (!validateTime(values.timeSpent)) {
    errors.fields.timeSpent = 'Formato de tempo incorreto'
  }

  if (!values.type) {
    errors.fields.type = 'Selecione um tipo'
  }

  if (!values.date) {
    errors.fields.date = 'Coloque uma data'
  }

  errors.error = (
    !!errors.fields.timeSpent
    || !!errors.fields.type
    || !!errors.fields.date
  )

  return errors
}
