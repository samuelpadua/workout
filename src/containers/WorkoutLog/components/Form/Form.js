import React, { useState } from 'react'
import PropTypes from 'prop-types'
import { Flex, Box } from '@rebass/grid'
import Input from '../../../../components/Input'
import Select from '../../../../components/Select'
import Button from '../../../../components/Button'
import useForm from './useForm'
import validate from './validadeForm'


function Form(props) {
  const {
    add
  } = props

  const {
    values,
    errors,
    handleChange,
    handleSubmit
  } = useForm(handleAddWorkout, validate); 

  function handleAddWorkout() {
    add(values)
  }

  const selectOptions = [
    {
      value: 'run',
      text: 'Run'
    },
    {
      value: 'swimming',
      text: 'Swimming'
    },
    {
      value: 'bike',
      text: 'Bike'
    }
  ]

  return (
    <form onSubmit={handleSubmit} noValidate>
      <Flex>
        <Box width={[ 1 ]} px={2}>
          <Input
            name="timeSpent"
            label="Tempo"
            value={values.timeSpent || ''}
            error={!!errors.fields.timeSpent}
            errorMessage={errors.fields.timeSpent}
            type="time"
            onChange={handleChange}
          />
        </Box>
        <Box width={[ 1 ]} px={2}>
          <Select
            name="type"
            label="Tipo"
            value={values.type || ''}
            error={!!errors.fields.type}
            errorMessage={errors.fields.type}
            options={selectOptions}
            onChange={handleChange}
          />
        </Box>
        <Box width={[ 1 ]} px={2}>
          <Input
            name="date"
            label="Data"
            value={values.date || ''}
            error={!!errors.fields.date}
            errorMessage={errors.fields.date}
            type="date"
            onChange={handleChange}
          />
        </Box>
        <Flex width={[ 1 ]} px={2} alignItems={errors.error ? 'center' : 'flex-end'}>
          <Button
            full
            type="submit"
          >
            Adicionar
          </Button>
        </Flex>
      </Flex>
    </form>
  )
}

Form.propTypes = {
  add: PropTypes.func.isRequired
}

export default Form
