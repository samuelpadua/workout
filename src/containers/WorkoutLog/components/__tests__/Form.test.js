import React from 'react'
import renderer from 'react-test-renderer'
import 'jest-styled-components'
import Form from '../Form/Form'

describe('Form', () => {
  it('should render Form to add new workout log', () => {
    const tree = renderer
      .create(<Form add={() => console.log('dispatch action to add new workout')} />)
      .toJSON()

    expect(tree).toMatchSnapshot()
  })
})
