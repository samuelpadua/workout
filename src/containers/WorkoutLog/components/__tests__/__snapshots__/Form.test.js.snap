// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form should render Form to add new workout log 1`] = `
.c1 {
  box-sizing: border-box;
  padding-left: 8px;
  padding-right: 8px;
  width: 100%;
}

.c3 {
  box-sizing: border-box;
  margin-bottom: 4px;
}

.c5 {
  box-sizing: border-box;
}

.c0 {
  box-sizing: border-box;
  padding-bottom: 20px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c2 {
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c9 {
  box-sizing: border-box;
  padding-left: 8px;
  padding-right: 8px;
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: flex-end;
  -webkit-box-align: flex-end;
  -ms-flex-align: flex-end;
  align-items: flex-end;
}

.c6 {
  position: relative;
  border-radius: 4px;
  box-sizing: border-box;
  border: 1px solid #8C8C8C;
  font: inherit;
  color: currentColor;
  width: 100%;
  height: 3em;
  margin: 0;
  padding: 6px 12px 7px;
}

.c6:focus {
  border-color: red;
  border: 2px solid #0086FF;
  padding: 5px 11px 6px;
  box-shadow: none !important;
  outline: 0;
}

.c4 {
  font-size: 14px;
  font-weight: 700;
  color: #404040;
}

.c8 {
  position: relative;
  border-radius: 4px;
  background: transparent;
  border: 1px solid #8C8C8C;
  font: inherit;
  color: currentColor;
  width: 100%;
  height: 3em;
  margin: 0;
  padding: 6px 12px 7px;
}

.c8:focus,
.c8:active {
  border-color: red;
  border: 2px solid #0086FF;
  padding: 5px 11px 6px;
  box-shadow: none !important;
  outline: 0;
  box-shadow: none !important;
}

.c7 {
  font-size: 14px;
  font-weight: 700;
  color: #404040;
}

.c10 {
  width: 100%;
  padding: 6px;
  border: 0;
  font: inherit;
  height: 3em;
  border-radius: 4px;
  background: #0086FF;
  box-shadow: none !important;
  outline: 0;
  color: white;
}

.c10:hover {
  background: #006de6;
}

.c10:active {
  background: #0053cc;
}

<form
  noValidate={true}
  onSubmit={[Function]}
>
  <div
    className="c0"
  >
    <div
      className="c1"
      width={
        Array [
          1,
        ]
      }
    >
      <div
        className="c2"
      >
        <div
          className="c3"
        >
          <label
            className="c4"
          >
            Tempo
          </label>
        </div>
        <div
          className="c5"
        >
          <input
            className="c6"
            label="Tempo"
            maxLength="5"
            name="timeSpent"
            onChange={[Function]}
            type="text"
            value=""
          />
        </div>
      </div>
    </div>
    <div
      className="c1"
      width={
        Array [
          1,
        ]
      }
    >
      <div
        className="c2"
      >
        <div
          className="c3"
        >
          <label
            className="c7"
          >
            Tipo
          </label>
        </div>
        <div
          className="c5"
        >
          <select
            className="c8"
            label="Tipo"
            name="type"
            onChange={[Function]}
            value=""
          >
            <option>
              Selecione
            </option>
            <option
              value="run"
            >
              Corrida
            </option>
            <option
              value="swimming"
            >
              Natação
            </option>
            <option
              value="bike"
            >
              Bicicleta
            </option>
          </select>
        </div>
      </div>
    </div>
    <div
      className="c1"
      width={
        Array [
          1,
        ]
      }
    >
      <div
        className="c2"
      >
        <div
          className="c3"
        >
          <label
            className="c4"
          >
            Data
          </label>
        </div>
        <div
          className="c5"
        >
          <input
            className="c6"
            label="Data"
            name="date"
            onChange={[Function]}
            type="date"
            value=""
          />
        </div>
      </div>
    </div>
    <div
      className="c9"
      width={
        Array [
          1,
        ]
      }
    >
      <button
        className="c10"
        color="default"
        type="submit"
      >
        Adicionar
      </button>
    </div>
  </div>
</form>
`;
